File: .devmind/scripts/check-graveyard.py
#!/usr/bin/env python3
"""
check-graveyard.py — Graveyard 关键词匹配检测
检测当前提议是否与已否决方案相似，防止重蹈覆辙

用法：
  python3 .devmind/scripts/check-graveyard.py "redis cache layer"
  python3 .devmind/scripts/check-graveyard.py "GraphQL API endpoint"

输出：
  匹配到的已否决方案列表，按匹配关键词数量降序排列
"""

import sys
from pathlib import Path


def load_graveyard(graveyard_dir: str) -> list[dict]:
    """加载所有 Graveyard 文件，提取关键词"""
    entries = []
    gdir = Path(graveyard_dir)
    if not gdir.exists():
        return entries

    for fpath in sorted(gdir.glob("*.md")):
        content = fpath.read_text(encoding="utf-8")
        lines = content.splitlines()

        # 提取标题
        title = next(
            (l.replace("## 放弃方案：", "").strip() for l in lines if l.startswith("## 放弃方案：")),
            fpath.stem
        )

        # 提取关键词行（格式：- 关键词：word1, word2, word3）
        keywords_line = next(
            (l for l in lines if l.startswith("- 关键词：")),
            ""
        )
        raw_keywords = keywords_line.replace("- 关键词：", "").strip()
        keywords = set(
            kw.strip().lower()
            for kw in raw_keywords.replace("，", ",").split(",")
            if kw.strip()
        )

        # 提取否决原因摘要
        reason_lines = []
        in_reason = False
        for l in lines:
            if l.startswith("- 放弃原因："):
                in_reason = True
                continue
            if in_reason:
                if l.startswith("  ") and l.strip():
                    reason_lines.append(l.strip())
                elif l.startswith("- ") and not l.startswith("  "):
                    break

        entries.append({
            "file": str(fpath),
            "title": title,
            "keywords": keywords,
            "reasons": reason_lines[:2],  # 只取前2条原因
        })

    return entries


def check_graveyard(proposal: str, graveyard_dir: str = ".devmind/memory/graveyard") -> list[dict]:
    """检测提议与已否决方案的关键词重叠"""
    # 将提议拆分为词（支持中英文混合）
    proposal_lower = proposal.lower()
    # 简单分词：按空格、逗号、句号分割
    import re
    proposal_terms = set(re.split(r"[\s,，。、]+", proposal_lower))
    proposal_terms.discard("")

    entries = load_graveyard(graveyard_dir)
    matches = []

    for entry in entries:
        if not entry["keywords"]:
            continue

        overlap = proposal_terms & entry["keywords"]
        # 也检查关键词是否是提议的子串（处理中文连续词）
        substring_matches = set()
        for kw in entry["keywords"]:
            if kw and kw in proposal_lower:
                substring_matches.add(kw)

        all_matches = overlap | substring_matches
        if all_matches:
            matches.append({
                "file": entry["file"],
                "title": entry["title"],
                "overlap": all_matches,
                "reasons": entry["reasons"],
                "match_count": len(all_matches),
            })

    return sorted(matches, key=lambda x: x["match_count"], reverse=True)


def main():
    if len(sys.argv) < 2:
        print("用法：python3 check-graveyard.py <提议描述>")
        print('示例：python3 check-graveyard.py "redis cache layer"')
        sys.exit(1)

    proposal = " ".join(sys.argv[1:])

    # 自动检测 graveyard 目录位置
    script_dir = Path(__file__).parent
    graveyard_dir = script_dir.parent / "memory" / "graveyard"

    matches = check_graveyard(proposal, str(graveyard_dir))

    if not matches:
        print(f'未发现与 "{proposal}" 相似的已否决方案。')
        sys.exit(0)

    print(f'⚠️  发现 {len(matches)} 个与提议相似的已否决方案：\n')
    for m in matches:
        print(f'  与已否决方案相似：{m["title"]}')
        print(f'  匹配关键词：{", ".join(sorted(m["overlap"]))}')
        if m["reasons"]:
            print(f'  否决原因：')
            for r in m["reasons"]:
                print(f'    {r}')
        print(f'  文件：{m["file"]}')
        print()

    sys.exit(1)  # 非零退出码，便于脚本调用时检测


if __name__ == "__main__":
    main()

File: .devmind/scripts/check-graveyard.js
#!/usr/bin/env node
// .devmind/scripts/check-graveyard.js
// Graveyard keyword matching detection
// Detects if current proposal is similar to rejected solutions
//
// Usage:
//   node .devmind/scripts/check-graveyard.js "redis cache layer"
//   node .devmind/scripts/check-graveyard.js "GraphQL API endpoint"
//
// Output:
//   List of matched rejected solutions, sorted by keyword overlap count

const fs = require('fs');
const path = require('path');

function loadGraveyard(graveyardDir) {
    const entries = [];

    if (!fs.existsSync(graveyardDir)) {
        return entries;
    }

    const files = fs.readdirSync(graveyardDir)
        .filter(f => f.endsWith('.md'))
        .sort();

    for (const filename of files) {
        const fpath = path.join(graveyardDir, filename);
        const content = fs.readFileSync(fpath, 'utf8');
        const lines = content.split('\n');

        // Extract title
        let title = path.basename(filename, '.md');
        for (const line of lines) {
            if (line.startsWith('## \u653E\u5F03\u65B9\u6848\uFF1A')) {
                title = line.replace('## \u653E\u5F03\u65B9\u6848\uFF1A', '').trim();
                break;
            }
        }

        // Extract keywords line
        let rawKeywords = '';
        for (const line of lines) {
            if (line.startsWith('- \u5173\u952E\u8BCD\uFF1A')) {
                rawKeywords = line.replace('- \u5173\u952E\u8BCD\uFF1A', '').trim();
                break;
            }
        }

        // Parse keywords (support both Chinese and English comma)
        const keywords = new Set(
            rawKeywords
                .replace(/\uFF0C/g, ',')
                .split(',')
                .map(kw => kw.trim().toLowerCase())
                .filter(kw => kw)
        );

        // Extract rejection reasons
        const reasonLines = [];
        let inReason = false;
        for (const line of lines) {
            if (line.startsWith('- \u653E\u5F03\u539F\u56E0\uFF1A')) {
                inReason = true;
                continue;
            }
            if (inReason) {
                if (line.startsWith('  ') && line.trim()) {
                    reasonLines.push(line.trim());
                } else if (line.startsWith('- ') && !line.startsWith('  ')) {
                    break;
                }
            }
        }

        entries.push({
            file: fpath,
            title,
            keywords,
            reasons: reasonLines.slice(0, 2) // Only first 2 reasons
        });
    }

    return entries;
}

function checkGraveyard(proposal, graveyardDir) {
    const proposalLower = proposal.toLowerCase();

    // Simple tokenization: split by spaces, commas, periods
    const proposalTerms = new Set(
        proposalLower
            .split(/[\s,\uFF0C\u3002\u3001]+/)
            .filter(t => t)
    );

    const entries = loadGraveyard(graveyardDir);
    const matches = [];

    for (const entry of entries) {
        if (entry.keywords.size === 0) continue;

        // Check word overlap
        const overlap = new Set();
        for (const term of proposalTerms) {
            if (entry.keywords.has(term)) {
                overlap.add(term);
            }
        }

        // Also check if keywords are substrings of proposal (for Chinese continuous words)
        for (const kw of entry.keywords) {
            if (kw && proposalLower.includes(kw)) {
                overlap.add(kw);
            }
        }

        if (overlap.size > 0) {
            matches.push({
                file: entry.file,
                title: entry.title,
                overlap: Array.from(overlap),
                reasons: entry.reasons,
                matchCount: overlap.size
            });
        }
    }

    // Sort by match count descending
    return matches.sort((a, b) => b.matchCount - a.matchCount);
}

function main() {
    if (process.argv.length < 3) {
        console.log('\u7528\u6CD5\uFF1Anode check-graveyard.js <\u63D0\u8BAE\u63CF\u8FF0>');
        console.log('\u793A\u4F8B\uFF1Anode check-graveyard.js "redis cache layer"');
        process.exit(1);
    }

    const proposal = process.argv.slice(2).join(' ');

    // Auto-detect graveyard directory location
    const scriptDir = __dirname;
    const graveyardDir = path.join(scriptDir, '..', 'memory', 'graveyard');

    const matches = checkGraveyard(proposal, graveyardDir);

    if (matches.length === 0) {
        console.log(`\u672A\u53D1\u73B0\u4E0E "${proposal}" \u76F8\u4F3C\u7684\u5DF2\u5426\u51B3\u65B9\u6848\u3002`);
        process.exit(0);
    }

    console.log(`\u26A0\uFE0F  \u53D1\u73B0 ${matches.length} \u4E2A\u4E0E\u63D0\u8BAE\u76F8\u4F3C\u7684\u5DF2\u5426\u51B3\u65B9\u6848\uFF1A\n`);

    for (const m of matches) {
        console.log(`  \u4E0E\u5DF2\u5426\u51B3\u65B9\u6848\u76F8\u4F3C\uFF1A${m.title}`);
        console.log(`  \u5339\u914D\u5173\u952E\u8BCD\uFF1A${m.overlap.sort().join(', ')}`);
        if (m.reasons.length > 0) {
            console.log('  \u5426\u51B3\u539F\u56E0\uFF1A');
            for (const r of m.reasons) {
                console.log(`    ${r}`);
            }
        }
        console.log(`  \u6587\u4EF6\uFF1A${m.file}`);
        console.log();
    }

    process.exit(1); // Non-zero exit code for script detection
}

main();
